### DISCUSSION TITLE
Add pgx_ulid extension - Increase Our Performance With ULID Type

### DISCUSSION THREAD

**Discussion by bitnom** [supabase/supabase#19499] [Feature Requests] [answered]
This feature request is for the [pgx_ulid](https://github.com/pksunkara/pgx_ulid) extension. I really like using UUID for primary keys but we all know the drawback of performance. This extension will mitigate a significant percentage of that.

There is a related [feature request for pg_idkit](https://github.com/orgs/supabase/discussions/9500) but it seems that extension would not provide nearly the same level of benefit (generally).

Please see below.

## From the README:

There are several different postgres extensions for [ulid][], but all of them have feature gaps. A good extension should have:

- **Generator**: A generator function to generate [ulid][] identifiers.
- **Binary**: Data be stored as binary and not text.
- **Type**: A postgres type `ulid` which is displayed as [ulid][] text.
- **Uuid**: Support for casting between UUID and [ulid][]
- **Timestamp**: Support to cast an [ulid][] to a timestamp
- **Monotonic**: Support [monotonicity][]

|                             Name                              | Language | Generator | Binary | Type |  UUID  | Timestamp | Monotonic |
| :-----------------------------------------------------------: | :------: | :-------: | :----: | :--: | :----: | :-------: | :-------: |
|      [`pgx_ulid`](https://github.com/pksunkara/pgx_ulid)      |   Rust   |    ✔️      |   ✔️    |  ✔️   |   ✔️    |    ✔️      |    ✔️      |
|       [`pgulid`](https://github.com/geckoboard/pgulid)        | PL/pgSQL |    ✔️      |   ❌   |  ❌  |   ❌   |    ❌     |    ❌     |
|      [`pg_idkit`](https://github.com/VADOSWARE/pg_idkit)      |   Rust   |    ✔️      |   ❌   |  ❌  |   ❌   |    ❌     |    ❌     |
|   [`uids-postgres`](https://github.com/spa5k/uids-postgres)   |   Rust   |    ✔️      | ⁉️[^1]  |  ❌  | ⁉️[^2]  |    ❌     |    ❌     |
|    [`pgsql_ulid`](https://github.com/scoville/pgsql-ulid)     | PL/pgSQL |    ❌     | ⁉️[^1]  |  ❌  |   ✔️    |    ❌     |    ❌     |
|        [`pg-ulid`](https://github.com/edoceo/pg-ulid)         |    C     |    ✔️      |   ❌   |  ❌  |   ❌   |    ❌     |    ❌     |
| [`ulid-postgres`](https://github.com/schinckel/ulid-postgres) | PL/pgSQL |    ✔️      |   ❌   |  ✔️   |   ❌   |    ✔️      |    ❌     |
|       [`pg_ulid`](https://github.com/iCyberon/pg_ulid)        |    Go    |    ✔️      |   ❌   |  ❌  |   ❌   |    ✔️      |    ❌     |
|        [`pg_ulid`](https://github.com/RPG-18/pg_ulid)         |   C++    |    ✔️      | ⁉️[^1]  |  ❌  |   ✔️    |    ❌     |    ❌     |

[^1]: You can convert the [ulid][] into `uuid` or `bytea` and store it like that.
[^2]: Supports casting indirectly through `bytea`.

## Why should I use ulid over uuid?

The main advantages are:

* Indexes created over ULIDs are less fragmented compared to UUIDs due to the timestamp and [monotonicity][] that was encoded in the ULID when it was created.
* ULIDs don't use special characters, so they can be used in URLs or even HTML.
* ULIDs are shorter than UUIDs as they are comprised of 26 characters compared to UUIDs' 36 characters.

This extension is approximately **30% faster** than both `pgcrypto`'s UUID and `pg_uuidv7`'s UUIDv7 when generating a million identifiers.

**Comment by FelixZY** [comment]
It might be possible to translate [ulid-mssql](https://github.com/rmalayter/ulid-mssql/blob/master/ulid.sql) to use a postgres dialect and get this running until the [more performant extension](https://github.com/pksunkara/pgx_ulid) is officially supported.

**Reply by bitnom** [reply]
you mean indexing a text field of it?

**Comment by erolabzait** [comment]
AWS Appsync has Ulid built in for DynamoDB, this will help the transition, i really want to try Supabase for new projects.

As a case against UUID watch this 2 videos:
- https://www.youtube.com/watch?v=a-K2C3sf1_Q
- https://www.youtube.com/watch?v=f53-Iw_5ucA

**Comment by inian** [comment, marked as answer]
Please upvote the extension in our Postgres extensions thread [here](https://github.com/orgs/supabase/discussions/679#discussioncomment-3822667). It helps us prioritise better among all the extension requests we receive
